Detectors/Raw/include/DetectorsRaw/RDHUtils.h

```diff
@@ -27,453 +28,721 @@ using LinkSubSpec_t = uint32_t;
 using IR = o2::InteractionRecord;
 
 struct RDHUtils {
+
+// disable is the type is a pointer
+#define NOTPTR(T) typename std::enable_if<!std::is_pointer<T>::value>::type* = 0
+// dereference SRC pointer as DST type reference
+#define TOREF(DST, SRC) *reinterpret_cast<DST*>(SRC)
+// dereference SRC pointer as DST type const reference
+#define TOCREF(DST, SRC) *reinterpret_cast<const DST*>(SRC)
+
   using RDHDef = o2::header::RAWDataHeader; // wathever is default
-  //using RDHv3 = o2::header::RAWDataHeaderV3; // V3 == V4
-  using RDHv4 = o2::header::RAWDataHeaderV4;
+  using RDHv4 = o2::header::RAWDataHeaderV4; // V3 == V4
   using RDHv5 = o2::header::RAWDataHeaderV5;
-  using RDHv6 = o2::header::RAWDataHeaderV6;
-  static constexpr uint8_t MaxRDHVersion = 6;
+  using RDHv6 = o2::header::RAWDataHeaderV6; // update this for every new version
+  using RDHLowest = RDHv4;                   // link this to lowest version
+  using RDHHighest = RDHv6;                  // link this to highest version
 
-  static constexpr int GBTWord = 16; // length of GBT word
-  static constexpr int MAXCRUPage = 512 * GBTWord;
+  /// make sure we RDH is a legitimate RAWDataHeader
+  template <typename RDH>
+  static void sanityCheckStrict()
+  {
+    static_assert(std::is_same<RDH, RDHv4>::value || std::is_same<RDH, RDHv5>::value ||
+                    std::is_same<RDH, RDHv6>::value,
+                  "not an RDH");
+  }
 
-  GPUhdi() static uint8_t getVersion(const RDHv4& rdh) { return rdh.version; } // same for all // why template does not work here?
-  GPUhdi() static uint8_t getVersion(const RDHv5& rdh) { return rdh.version; } // same for all
-  GPUhdi() static uint8_t getVersion(const RDHv6& rdh) { return rdh.version; } // same for all
-  GPUhdi() static uint8_t getVersion(const void* rdhP)
+  /// make sure we RDH is a legitimate RAWDataHeader or generic RDHAny placeholder
+  template <typename RDH>
+  static void sanityCheckLoose()
+  {
+    static_assert(std::is_same<RDH, RDHv4>::value || std::is_same<RDH, RDHv5>::value ||
+                    std::is_same<RDH, RDHv6>::value || std::is_same<RDHAny, RDH>::value,
+                  "not an RDH or RDHAny");
+  }
+
+  /// get numeric version of the RDH
+  template <typename RDH>
+  static constexpr int getVersion()
   {
-    auto v = getVersion((*reinterpret_cast<const RDHDef*>(rdhP)));
-    if (v > MaxRDHVersion) {
-      processError(v, "version");
+    sanityCheckStrict<RDH>();
+    if (std::is_same<RDH, RDHv6>::value) {
+      return 6;
+    }
+    if (std::is_same<RDH, RDHv5>::value) {
+      return 5;
+    }
+    if (std::is_same<RDH, RDHv4>::value) {
+      return 4;
     }
-    return v;
   }
+
+  ///=====================================================
+  /// placeholder class for arbitraty-version 64B-lonh RDH
+  struct RDHAny {
```

Comment:
move to separate header file in DataFormats/Headers

Reply:
I want to have this class with root dictionary, is it ok for DataFormats/Headers? 
Also note that in this case others (e.g. DataDistribution) also will be exposed to it.

Reply:
I see, so we need to add this in the cmake setup for Headers, we don't generate dictionaries there yet. I do not see a problem, neither for exposing it. Its public also in Detectors/Raw. We can also move it later.
Are you storing RDHAny with ROOT I/O? Does it make sense to have this structure of 64bit words in a Root file?